### PR TITLE
bi copy support cross platform

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

File: src/socket.rs [hunk 1]
```diff
@@ -166,3 +168,31 @@ mod linux {
...
+const EINVAL: i32 = 22;
+pub async fn relay(
+    downstream: &mut tokio::net::TcpStream,
+    upstream: &mut tokio::net::TcpStream,
+) -> Result<(), Error> {
+    #[cfg(all(target_os = "linux"))]
+    {
+        match realm_io::bidi_zero_copy(downstream, upstream).await {
+            Ok(()) => Ok(()),
+            Err(ref e) if e.raw_os_error().map_or(false, |ec| ec == EINVAL) => {
```
Comment (hzxuzhonghu) [inline src/socket.rs:181]: This is referred from https://github.com/zhboner/realm/blob/master/realm_core/src/tcp/plain.rs

File: src/socket.rs [hunk 2]
```diff
@@ -166,3 +168,31 @@ mod linux {
...
+                    Err(e) => Err(e),
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    {
+        match tokio::io::copy_bidirectional(downstream, upstream).await {
```
Comment (howardjohn) [inline src/socket.rs:193]: I think you can replace this with `map(|_| ())`

**Review by howardjohn — COMMENTED** [review]
lgtm just 1 question

**Review by howardjohn — APPROVED** [review]
oops forgot to ask the questions but you answered it already...

File: src/socket.rs (hunk 1 above)
Comment (howardjohn) [reply src/socket.rs:181]: any idea when it can occur on Linux?

Comment (hzxuzhonghu) [reply src/socket.rs:181]:
```
       EINVAL The target filesystem doesn't support splicing.

       EINVAL The target file is opened in append mode.

       EINVAL Neither of the file descriptors refers to a pipe.

       EINVAL An offset was given for nonseekable device (e.g., a pipe).

       EINVAL fd_in and fd_out refer to the same pipe.

```

Comment (hzxuzhonghu) [reply src/socket.rs:181]: this is from `man splice`

Comment (howardjohn) [reply src/socket.rs:181]: hmm.. are any of those even possible for our usage?

File: src/socket.rs (hunk 2 above)
Comment (hzxuzhonghu) [reply src/socket.rs:193]: good suggesttion

File: src/socket.rs (hunk 1 above)
Comment (hzxuzhonghu) [reply src/socket.rs:181]: I am not sure, just in case

**Pull request merged**